commit 7001052160d172f6de06adeffde24dde9935ece8
Merge: f022814633e1 3986f65d4f40
Author: Linus Torvalds <torvalds@linux-foundation.org>
Date:   Sun Mar 27 10:17:23 2022 -0700

    Merge tag 'x86_core_for_5.18_rc1' of git://git.kernel.org/pub/scm/linux/kernel/git/tip/tip
    
    Pull x86 CET-IBT (Control-Flow-Integrity) support from Peter Zijlstra:
     "Add support for Intel CET-IBT, available since Tigerlake (11th gen),
      which is a coarse grained, hardware based, forward edge
      Control-Flow-Integrity mechanism where any indirect CALL/JMP must
      target an ENDBR instruction or suffer #CP.
    
      Additionally, since Alderlake (12th gen)/Sapphire-Rapids, speculation
      is limited to 2 instructions (and typically fewer) on branch targets
      not starting with ENDBR. CET-IBT also limits speculation of the next
      sequential instruction after the indirect CALL/JMP [1].
    
      CET-IBT is fundamentally incompatible with retpolines, but provides,
      as described above, speculation limits itself"
    
    [1] https://www.intel.com/content/www/us/en/developer/articles/technical/software-security-guidance/technical-documentation/branch-history-injection.html
    
    * tag 'x86_core_for_5.18_rc1' of git://git.kernel.org/pub/scm/linux/kernel/git/tip/tip: (53 commits)
      kvm/emulate: Fix SETcc emulation for ENDBR
      x86/Kconfig: Only allow CONFIG_X86_KERNEL_IBT with ld.lld >= 14.0.0
      x86/Kconfig: Only enable CONFIG_CC_HAS_IBT for clang >= 14.0.0
      kbuild: Fixup the IBT kbuild changes
      x86/Kconfig: Do not allow CONFIG_X86_X32_ABI=y with llvm-objcopy
      x86: Remove toolchain check for X32 ABI capability
      x86/alternative: Use .ibt_endbr_seal to seal indirect calls
      objtool: Find unused ENDBR instructions
      objtool: Validate IBT assumptions
      objtool: Add IBT/ENDBR decoding
      objtool: Read the NOENDBR annotation
      x86: Annotate idtentry_df()
      x86,objtool: Move the ASM_REACHABLE annotation to objtool.h
      x86: Annotate call_on_stack()
      objtool: Rework ASM_REACHABLE
      x86: Mark __invalid_creds() __noreturn
      exit: Mark do_group_exit() __noreturn
      x86: Mark stop_this_cpu() __noreturn
      objtool: Ignore extra-symbol code
      objtool: Rename --duplicate to --lto
      ...

diff --cc arch/x86/net/bpf_jit_comp.c
index 6efbb87f65ed,b592ea0fc150..8fe35ed11fd6
--- a/arch/x86/net/bpf_jit_comp.c
+++ b/arch/x86/net/bpf_jit_comp.c
@@@ -380,7 -395,14 +391,14 @@@ int bpf_arch_text_poke(void *ip, enum b
  		/* BPF poking in modules is not supported */
  		return -EINVAL;
  
+ 	/*
+ 	 * See emit_prologue(), for IBT builds the trampoline hook is preceded
+ 	 * with an ENDBR instruction.
+ 	 */
+ 	if (is_endbr(*(u32 *)ip))
+ 		ip += ENDBR_INSN_SIZE;
+ 
 -	return __bpf_arch_text_poke(ip, t, old_addr, new_addr, true);
 +	return __bpf_arch_text_poke(ip, t, old_addr, new_addr);
  }
  
  #define EMIT_LFENCE()	EMIT3(0x0F, 0xAE, 0xE8)